pkg/storage/etcd.go

```diff
@@ -36,37 +37,75 @@ func NewEtcdStorage(endpoints []string) (*EtcdStorage, error) {
 }
 
 // Store stores the key
-func (etcds *EtcdStorage) Store(key string, value Data) error {
+func (etcds *EtcdStorage) StoreCheckin(key string, value time.Time) error {
 	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
-	encoded, _ := json.Marshal(value)
-	_, err := etcds.kv.Put(ctx, fmt.Sprintf("env=%s", key), string(encoded))
-	cancel()
+	defer cancel()
+	encoded, err := json.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "failed to store key")
+	}
+	key = strings.TrimSuffix(strings.TrimPrefix(key, "env/"), "/last") // sanitize the strings
+	_, err = etcds.kv.Put(ctx, fmt.Sprintf("env/%s/last", key), string(encoded))
+	return errors.Wrap(err, "failed to store key")
+}
+
+func (etcds *EtcdStorage) StoreIncident(key string, date string, value []time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	encoded, err := json.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "failed to store key")
+	}
+	key = strings.TrimSuffix(strings.TrimPrefix(key, "env/"), "/last") // sanitize the strings
+	_, err = etcds.kv.Put(ctx, fmt.Sprintf("env/%s/incidents/%s", key, date), string(encoded))
 	return errors.Wrap(err, "failed to store key")
 }
 
 // All returns a map of all keys and values
-func (etcds *EtcdStorage) All() (map[string]Data, error) {
+func (etcds *EtcdStorage) AllCheckins() (map[string]time.Time, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	v, err := etcds.kv.Get(ctx, "env", clientv3.WithPrefix())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve all checkins")
+	}
+	all := map[string]time.Time{}
+	for _, item := range v.Kvs {
+		if strings.Contains(string(item.Key), "last") {
+			var data time.Time
+			if err := json.Unmarshal(item.Value, &data); err != nil {
+				return nil, errors.Wrap(err, "failed to retrieve all checkins")
+			}
+			all[string(item.Key)] = data
+		}
+	}
+	return all, nil
+}
+
+func (etcds *EtcdStorage) AllIncidents() (map[string][]string, error) {
```

Comment:
Here it would be clearer to return a more clearly typed result.

e.g. `map[string][]Incident`